Add register service method for DTD

helin24 opened the pull request:
This is for enabling the Flutter plugin to register a service method for navigating to code, something that can be requested by a user from Flutter DevTools. More info on service methods: https://github.com/dart-lang/sdk/blob/main/pkg/dtd_impl/dtd_protocol.md#service-methods

Example DTD interactions in logs (comments for clarifying two-client interaction):
```
INFO - #c.j.l.d.i.d.DartDevToolsService - Starting Dart DevTools, sdk 3.5.0-321.0.dev
INFO - #c.j.l.d.i.t.DartToolingDaemonService - Connected to DTD successfully
FINE - #c.j.l.d.i.t.DartToolingDaemonService - --> {"jsonrpc":"2.0","method":"FileSystem.setIDEWorkspaceRoots","id":"1","params":{"roots":["file:///Users/helinx/IdeaProjects/hello_flutter_external/android","file:///Users/helinx/IdeaProjects/hello_flutter_external"],"secret":"qzgLfkLrkpov0FXC"}}
// client 1 registers service
FINE - #c.j.l.d.i.t.DartToolingDaemonService - --> {"jsonrpc":"2.0","method":"registerService","id":"2","params":{"service":"Test","method":"testMethod"}}
FINE - #c.j.l.d.i.t.DartToolingDaemonService - <-- {"jsonrpc":"2.0","result":{"type":"Success"},"id":"1"}
FINE - #c.j.l.d.i.t.DartToolingDaemonService - <-- {"jsonrpc":"2.0","result":{"type":"Success"},"id":"2"}

// client 2 calls service method
FINE - #c.j.l.d.i.t.DartToolingDaemonService - --> {"jsonrpc":"2.0","method":"Test.testMethod","id":"3","params":{"param1":"1"}}

// client 1 receives service method request
FINE - #c.j.l.d.i.t.DartToolingDaemonService - <-- {"jsonrpc":"2.0","method":"Test.testMethod","id":0,"params":{"param1":"1"}}

// client 1 sends service method response
FINE - #c.j.l.d.i.t.DartToolingDaemonService - --> {"jsonrpc":"2.0","id":"0","result":{"success":true}}

// client 2 receives response
FINE - #c.j.l.d.i.t.DartToolingDaemonService - <-- {"jsonrpc":"2.0","result":{"success":true},"id":"3"}
```

helin24 commented:
I'm not sure how the response gets sent back to DTD yet - so this is just an early draft for @alexander-doroshko to leave comments on (whenever convenient).

alexander-doroshko commented:
LGTM. Let me know when it's production-ready. 

Note that I gonna be offline between July 20 and August 11.

helin24 commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -237,6 +272,21 @@ class DartToolingDaemonService private constructor(private val project: Project)
       //  println(response)
       //}
+
+      // Example request to test registering a service method
```
This part should be commented out and only used if we want to test this out

alexander-doroshko replied:
Ok, pushed as commented out

helin24 commented:
@alexander-doroshko I fixed this up and it's production ready now. I added an example of logs to the original comment, let me know if you have any questions!

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -105,6 +106,24 @@ class DartToolingDaemonService private constructor(private val project: Project)
+    sendRequest("registerService", params, false) {response ->
+      val result = response.getAsJsonObject("result")
+      if (result == null) {
+        logger.error("No result from attempt to register service $service.$method", service, method)
```
I removed the second and the third arguments because both `$service` and `$method` are already in the message (the first argument)

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -105,6 +106,24 @@ class DartToolingDaemonService private constructor(private val project: Project)
+      val result = response.getAsJsonObject("result")
+      if (result == null) {
+        logger.error("No result from attempt to register service $service.$method", service, method)
+      }
```
Should there be `return@sendRequest` right after logging an error?

helin24 replied:
Yes, it would make sense to return after the error.

alexander-doroshko replied:
done

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -105,6 +106,24 @@ class DartToolingDaemonService private constructor(private val project: Project)
+      }
+      val type = result.getAsJsonPrimitive("type")
+      if (type == null || "Success" != type.asString) {
+        logger.error("Failed to register service $service.$method", service, method)
```
The same. I removed the second and the third arguments because both $service and $method are already in the message (the first argument)

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -105,6 +106,24 @@ class DartToolingDaemonService private constructor(private val project: Project)
+      val type = result.getAsJsonPrimitive("type")
+      if (type == null || "Success" != type.asString) {
+        logger.error("Failed to register service $service.$method", service, method)
+      }
```
The same. Should there be return@sendRequest right after logging an error?

helin24 replied:
Yes, same here!

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -128,6 +147,22 @@ class DartToolingDaemonService private constructor(private val project: Project)
     webSocket.send(requestString)
   }
 
+  fun sendResponse(id: String, result: JsonObject) {
```
I made this function private. It is ok?

helin24 replied:
Yes it should only be used in this class

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt:
```diff
@@ -256,6 +306,16 @@ class DartToolingDaemonService private constructor(private val project: Project)
+        val params = json["params"].asJsonObject
+        val id = json["id"].asString
+        val consumer = servicesMap.get(method)
+        val result = consumer?.received(params)
+        if (result == null) {
+          logger.warn("No response from request consumer for message: $text")
+        } else {
+          sendResponse(id, result)
+        }
```
I rewrote it in a way that neither `consumer`, nor `result` are nullable inside the `else if` branch:
```
      val method = json["method"]?.asString
      val serviceConsumer = servicesMap[method]
      if (method == "streamNotify") {...}
      else if (serviceConsumer != null) {
        val params = json["params"].asJsonObject
        val id = json["id"].asString
        val result = serviceConsumer.received(params)
        sendResponse(id, result)
      }
      else {...}
```
Is it ok?

helin24 replied:
Looks great! Thanks

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonServiceConsumer.kt:
```diff
@@ -0,0 +1,8 @@
+import com.google.gson.JsonObject
+
+fun interface DartToolingDaemonServiceConsumer {
+  fun received(request: JsonObject): JsonObject
```
Maybe rename `received` to something like `handleRequest`?

The implementations can't fail and always return valid not null `JsonObject`, right?

helin24 replied:
1. `handleRequest` sounds better! change to `RequestHandler` class name also fine
2. Yes, if there is an error the resulting `JsonObject` should reflect that.

alexander-doroshko replied:
done

alexander-doroshko commented:
Pushed, thanks. Please have a look at the comments.

helin24 commented:
> One more thought. Can it become a problem that `DartToolingDaemonServiceConsumer#received` is called synchronously in `DtdWebSocketEventHandler#onMessage`?
> 
> As I understand, messages processing by DtdWebSocketEventHandler is blocked until `DartToolingDaemonServiceConsumer` responds. OTOH, if all `DartToolingDaemonServiceConsumer` impolementations are going to respond instantly, it shouldn't cause issues.

Hmmm... good point. I don't think we should assume that all implementations would respond instantly. What are implications for making this async? Sending the response to DTD later should be reasonable.

alexander-doroshko commented:
> Sending the response to DTD later should be reasonable.

Ok, I moved potentially long calculations to a different thread:
```
      else if (serviceConsumer != null) {
        val params = json["params"].asJsonObject
        val id = json["id"].asString
        ApplicationManager.getApplication().executeOnPooledThread {
          val result = serviceConsumer.handleRequest(params)
          sendResponse(id, result)
        }
      }
```

I didn't test but I hope it works.

BTW, we may want to review the whole class implementation, how ready it is for concurrent multi-threaded usage

alexander-doroshko commented:
> how to test this

I'm not sure about real testing. But we can carefully analyze all the usages of the shared data (DartToolingDaemonService class fields) and think what would happen if many parallel threads executed the same (where applicable) or different functions of this class and read/wrote the shared data in the most unlucky moments
